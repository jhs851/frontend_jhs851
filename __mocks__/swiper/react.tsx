import React from 'react'

const mock: React.FC = ({ children }) => <div>{children}</div>

export default mock
