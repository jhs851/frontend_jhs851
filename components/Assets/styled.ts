import styled from 'styled-components'

export const Container = styled.div`
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  z-index: 1030;
  background-color: ${({ theme }) => theme.primary};
`

export const Wrapper = styled.div`
  padding: 11px;
`

export const Asset = styled.button`
  height: 46px;
  border-radius: 5px;
  overflow: hidden;
`
