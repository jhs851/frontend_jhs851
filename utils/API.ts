import { Asset, Message, Room, User, VirtualMessage } from 'types'
import {
  assets as mockAssets,
  messages as mockMessages,
  rooms as mockRooms,
  users as mockUsers,
} from 'utils/mocks'

export default class API {
  static lastMessageID: number = mockMessages.slice(-1)[0].id
  static users: User[] = mockUsers
  static rooms: Room[] = mockRooms
  static messages: Message[] = mockMessages
  static assets: Asset[] = mockAssets

  static send<T>(
    executor: (
      resolve: (value: T | PromiseLike<T>) => void,
      reject: (reason?: any) => void,
    ) => void,
  ): Promise<T> {
    return new Promise((resolve, reject) => {
      setTimeout(() => executor(resolve, reject), 600)
    })
  }

  static async findUser(user_id: number): Promise<User> {
    return API.send((resolve, reject) => {
      const user = API.users.find(user => user.id === user_id)

      if (!user) {
        return reject(`Not found User. user_id: ${user_id}`)
      }

      resolve(user)
    })
  }

  static async getRooms(): Promise<Room[]> {
    return API.send(resolve => resolve(API.rooms))
  }

  static async findRoom(room_id: number): Promise<Room> {
    return API.send((resolve, reject) => {
      const room = API.rooms.find(room => room.id === room_id)

      if (!room) {
        return reject(`Not found Room. room_id: ${room_id}`)
      }

      resolve(room)
    })
  }

  static async getMessages(room_id: number): Promise<Message[]> {
    return API.send(resolve =>
      resolve(API.messages.filter(message => message.room_id === room_id)),
    )
  }

  static async findMessage(message_id: number): Promise<Message> {
    return API.send((resolve, reject) => {
      const message = API.messages.find(message => message.id === message_id)

      if (!message) {
        return reject(`Not found User. message_id: ${message_id}`)
      }

      resolve(message)
    })
  }

  static async getAssets(): Promise<Asset[]> {
    return API.send(resolve => resolve(API.assets))
  }

  static async markReadRoom(room_id: number): Promise<void> {
    return API.send(resolve => {
      API.rooms = API.rooms.map(room =>
        room.id === room_id ? { ...room, unread_message_count: 0 } : room,
      )
      resolve()
    })
  }

  static async updateRoomLastMessage(
    room_id: number,
    last_message_id: number,
  ): Promise<void> {
    return API.send(resolve => {
      API.rooms = API.rooms.map(room =>
        room.id === room_id ? { ...room, last_message_id } : room,
      )
      resolve()
    })
  }

  static async createMessage(data: VirtualMessage): Promise<Message> {
    return API.send(async resolve => {
      const id = ++API.lastMessageID
      const message = {
        ...data,
        id,
      }
      API.messages = API.messages.concat(message)
      await API.updateRoomLastMessage(data.room_id, id)

      resolve(message)
    })
  }
}
