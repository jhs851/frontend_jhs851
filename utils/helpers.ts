import { RelatedMessage as RMessage, VirtualMessage as VMessage } from 'types'

export const isVirtualMessage = (
  message: RMessage | VMessage,
): message is VMessage => (message as VMessage).virtual_id !== undefined
