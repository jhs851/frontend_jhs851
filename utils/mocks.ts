import { Asset, Message, Room, User } from 'types'
import moment from 'moment'

export const users: User[] = [
  {
    id: 1,
    name: '지그재그',
    avatar: '',
  },
  {
    id: 2,
    name: '장만월',
    avatar: '/avatars/avatar1.png',
  },
  {
    id: 3,
    name: '신정근',
    avatar: '/avatars/avatar3.png',
  },
  {
    id: 4,
    name: '이미라',
    avatar: '/avatars/avatar4.png',
  },
  {
    id: 5,
    name: '구찬성',
    avatar: '/avatars/avatar5.png',
  },
  {
    id: 6,
    name: '노준석',
    avatar: '/avatars/avatar6.png',
  },
  {
    id: 7,
    name: '김유나',
    avatar: '/avatars/avatar7.png',
  },
  {
    id: 8,
    name: '구현모',
    avatar: '/avatars/avatar2.png',
  },
]

export const rooms: Room[] = [
  {
    id: 1,
    user_id: 2,
    local_name: '장만월 사장님',
    unread_message_count: 2,
    last_message_id: 4,
  },
  {
    id: 2,
    user_id: 3,
    local_name: '신정근 바텐더',
    unread_message_count: 0,
    last_message_id: 5,
  },
  {
    id: 3,
    user_id: 4,
    local_name: '이미라 의사',
    unread_message_count: 0,
    last_message_id: 6,
  },
  {
    id: 4,
    user_id: 5,
    local_name: '구찬성 지배인',
    unread_message_count: 0,
    last_message_id: 7,
  },
  {
    id: 5,
    user_id: 6,
    local_name: '노준석 총지배인',
    unread_message_count: 0,
    last_message_id: 8,
  },
  {
    id: 6,
    user_id: 7,
    local_name: '김유나 인턴',
    unread_message_count: 0,
    last_message_id: 9,
  },
  {
    id: 7,
    user_id: 8,
    local_name: '구현모',
    unread_message_count: 0,
    last_message_id: 10,
  },
]

export const messages: Message[] = [
  {
    id: 1,
    user_id: 2,
    room_id: 1,
    message: '출근했니?',
    created_at: moment('2020-05-06 02:34:00').toISOString(),
  },
  {
    id: 2,
    user_id: 2,
    room_id: 1,
    message: '출근했냐구?',
    created_at: moment('2020-05-06 02:35:00').toISOString(),
  },
  {
    id: 3,
    user_id: 2,
    room_id: 1,
    message: '어딘데 출근 안하니?',
    created_at: moment('2020-05-07 02:36:00').toISOString(),
  },
  {
    id: 4,
    user_id: 2,
    room_id: 1,
    message: '어딘데 출근 안하니, 죽고싶니?',
    created_at: moment().set({ h: 9, m: 32 }).toISOString(),
  },
  {
    id: 5,
    user_id: 3,
    room_id: 2,
    message: '오시는 길에 와인 몇병만 사다주세요.',
    created_at: moment().set({ h: 2, m: 34 }).toISOString(),
  },
  {
    id: 6,
    user_id: 4,
    room_id: 3,
    message:
      '휴가 잘 보내고 계신가요? 다름이 아니라 지금 검사 결과가 나와서 연락드립니다.',
    created_at: moment().subtract(1, 'day').toISOString(),
  },
  {
    id: 7,
    user_id: 5,
    room_id: 4,
    message: '아 휴가셨군요. 약속은 다음으로 미루시죠!',
    created_at: moment().subtract(2, 'day').toISOString(),
  },
  {
    id: 8,
    user_id: 6,
    room_id: 5,
    message: '휴가에서 언제 돌아오시는지요. 돌아오시면 회포 한번 푸시죠.',
    created_at: moment().subtract(3, 'day').toISOString(),
  },
  {
    id: 9,
    user_id: 7,
    room_id: 6,
    message: '304호 키를 잃어버렸어요 어떻게 해야하죠 ㅠ',
    created_at: moment().subtract(4, 'day').toISOString(),
  },
  {
    id: 10,
    user_id: 8,
    room_id: 7,
    message: '술먹자',
    created_at: moment().subtract(5, 'day').toISOString(),
  },
]

export const assets: Asset[] = [
  {
    id: 1,
    path: '/assets/1.png',
  },
  {
    id: 2,
    path: '/assets/2.png',
  },
  {
    id: 3,
    path: '/assets/3.png',
  },
  {
    id: 4,
    path: '/assets/4.png',
  },
  {
    id: 5,
    path: '/assets/5.png',
  },
  {
    id: 6,
    path: '/assets/6.png',
  },
  {
    id: 7,
    path: '/assets/7.png',
  },
]
