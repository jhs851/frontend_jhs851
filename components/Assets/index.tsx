import React, { useRef, useState } from 'react'
import { Asset, Container, Wrapper } from './styled'
import Image from 'components/Image'
import { Asset as IAsset } from 'types'
import { Swiper, SwiperSlide } from 'swiper/react'

export type HTMLDivAssetsElement =
  | (HTMLDivElement & {
      onToggle: () => void
    })
  | undefined

type P = {
  assets: IAsset[]
  onClick: (asset: string) => any
}

const Assets: React.ForwardRefRenderFunction<HTMLDivAssetsElement, P> = (
  { assets, onClick },
  ref,
) => {
  const containerRef = useRef<HTMLDivElement | null>(null)
  const [visible, setVisible] = useState<boolean>(false)
  const onToggle = () => {
    if (!assets.length) {
      return
    }

    setVisible(prevState => !prevState)
  }

  React.useImperativeHandle(ref, () => {
    if (containerRef.current) {
      return {
        ...containerRef.current,
        onToggle,
      }
    }

    return undefined
  })

  if (!visible) {
    return null
  }

  return (
    <Container ref={containerRef}>
      <Wrapper>
        <Swiper spaceBetween={10} slidesPerView="auto">
          {assets.map(asset => (
            <SwiperSlide key={asset.id} style={{ width: 46 }}>
              <Asset onClick={() => onClick(asset.path)}>
                <Image src={asset.path} width={46} height={46} alt="이미지" />
              </Asset>
            </SwiperSlide>
          ))}
        </Swiper>
      </Wrapper>
    </Container>
  )
}

export default React.forwardRef(Assets)
