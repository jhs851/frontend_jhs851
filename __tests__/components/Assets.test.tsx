import { render } from '@testing-library/react'
import Assets from 'components/Assets'
import { assets as mockAssets } from 'utils/mocks'

describe('Assets 컴포넌트', () => {
  it('렌더링이 올바르게 된다', () => {
    const assets = mockAssets
    const onClick = jest.fn()

    expect(
      render(<Assets assets={assets} onClick={onClick} />).container,
    ).toMatchSnapshot()
  })
})
