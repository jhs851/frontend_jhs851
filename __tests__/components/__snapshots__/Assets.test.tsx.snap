// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Assets 컴포넌트 렌더링이 올바르게 된다 1`] = `<div />`;
